### ISSUE TITLE
Update `ConfigureHttpClientDefaults` to provide `IHttpClientBuilder` with the actual `Name` of an `HttpClient`

### ISSUE THREAD

**Issue by iliar-turdushev** [dotnet/runtime#101719]
## Problem

[ConfigureHttpClientDefaults](https://github.com/dotnet/runtime/blame/2ad47cd52ebe57c1b4c2e28f95cc4eed6fcb354d/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientFactoryServiceCollectionExtensions.cs#L80C9-L80C141) allows to set up default configuration for `HttpClient`s. A delegate of type `Action<IHttpClientBuilder>` which is passed to this method configures the defaults. The `IHttpClientBuilder` argument has a property `Name` which is the name of named `HttpClient`. But `ConfigureHttpClientDefaults` always sets `Name` to `null`. Sometimes it could be beneficial to set up defaults per name of `HttpClient`.

## Question

Can we update `ConfigureHttpClientDefaults` to provide the actual `Name` of named `HttpClient`?

## Known issues

Such behavior causes issues with [AddStandardResiliencePipeline](https://github.com/dotnet/extensions/blob/98eac54bb554dbf724136c20543f6cbff0663a3c/src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHttpClientBuilderExtensions.StandardResilience.cs#L65C5-L65C119) functionality from [dotnet/extensions](https://github.com/dotnet/extensions). `AddStandardResiliencePipeline` relies on `IHttpClientBuilder.Name` to register a resilience pipeline and then bind it to the corresponding named `HttpClient`. But since `ConfigureHttpClientDefaults` always sets `IHttpClientBuilder.Name` to `null` we unable to use it to configure the standard resilience pipeline per each named `HttpClient`. For example, the following two samples produce different results because `AddStandardResiliencePipeline` relies on `IHttpClientBuilder.Name`:

Configuration 1
```csharp
// Registers a single resilience pipeline named "standard", because IHttpClientBuilder.Name = null
serviceCollection.ConfigureHttpClientDefaults(x => x.AddStandardResiliencePipeline());

var factory = serviceCollection.BuildServiceProvider().GetRequiredService<IHttpClientFactory>();

// Both "one" and "two" use common "standard" pipeline and share its state
var one = factory.CreateClient("one");
var two = factory.CreateClient("two");
```

Configuration 2
```csharp
// Registers two resilience pipelines named "one-standard" and "two-standard"
serviceCollection.AddHttpClient("one").AddStandardResiliencePipeline();
serviceCollection.AddHttpClient("two").AddStandardResiliencePipeline();

var factory = serviceCollection.BuildServiceProvider().GetRequiredService<IHttpClientFactory>();

// "one" and "two" has its own resilience pipeline
var one = factory.CreateClient("one");
var two = factory.CreateClient("two");
```

**Comment by CarnaViire** [comment]
Thanks for the detailed report and examples @iliar-turdushev!
On one hand, I can see the value in being able to access the name during the callback execution.
On the other hand, sharing the state/instance of the default resilience pipeline looks a bit like a lifetime configuration problem, i.e. the pipeline registration. If the object/service is stateful, I believe it's better to avoid it being an essential (named) singleton (unless it is required by design, and I don't think the sharing was intended in principle, was it?)

**Comment by iliar-turdushev** [comment]
> If the object/service is stateful, I believe it's better to avoid it being an essential (named) singleton (unless it is required by design, and I don't think the sharing was intended in principle, was it?)

I'm not an expert in dotnet/extensions Resilience, but, as I understand, it is by design. All instances of a particular named `HttpClient` should share a single resilience pipeline. It allows us to enforce such stateful resilience policies as `RateLimiter` and `CircuitBreaker`. For example, given named `HttpClient` should not exceed some limit of concurrent HTTP requests.

@martintmk Could you, please, confirm that it is by design to share resilience pipelines per named `HttpClient`? Thanks.

**Comment by iliar-turdushev** [comment]
@CarnaViire Can we follow up on this? Who can help us? Thanks.

**Comment by CarnaViire** [comment]
@iliar-turdushev I have it on my list. Sorry for the delay. I will give you an estimate and more extended answer next week.

**Comment by CarnaViire** [comment]
@iliar-turdushev I am sorry that due to the load it took so long for me to get to it -- only to realize I've misunderstood the proposal and the problem that stays behind it. 😰

Short answer will be: _unfortunately, it's impossible to provide the name_.

Simply because the exact name _cannot_ be known at that moment -- the moment being _at the time of registration_.

This is the way it works by design. At the moment of registration, there's no knowledge of a specific name. `null` serves as "any name" sign, similar to e.g. `KeyedService.AnyKey`. This is also the way Options pattern work by design, see the docs for [Named options support using IConfigureNamedOptions](https://learn.microsoft.com/en-us/dotnet/core/extensions/options#named-options-support-using-iconfigurenamedoptions)

> The `null` named option is used to target all of the named instances instead of a specific named instance. [ConfigureAll](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.optionsservicecollectionextensions.configureall) and [PostConfigureAll](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.optionsservicecollectionextensions.postconfigureall) use this convention.

The name _will_ be known, but only at the moment of _execution_ -- Options infra applying IConfigureOptions callbacks to get the value of the specific named option -- when you actually create a specific handler and other configuration callbacks are called (like the ones passed to ConfigurePrimaryHttpMessageHandler).

Basically, the default stuff is _registered_ only once, but then _executed_ multiple times for every named option.

***

This is something that should be solved at the level of integration with resilience pipelines. In a way so `null` pipeline will become a "transient" instead of a "singleton". I will spend today trying to investigate possible solutions, but my Polly knowledge is limited. Anyway, I will let you know what I find today or by Monday morning. After that I'll have to close this issue as "By design".

Please let me know if I missed or misunderstood something 🥺

**Comment by CarnaViire** [comment]
@iliar-turdushev Thinking more about this:

1) First problem is the registration of a ResiliencePipeline by the name/key. I believe that the following can be utilized to defer the registration until later: https://github.com/App-vNext/Polly/blob/main/docs/advanced/dependency-injection.md#deferred-addition-of-pipelines -- I guess smth like

```c#
services.AddResiliencePipelines<TKey, TValue>(ctx =>
{
    foreach (var clientName in Helper.GetAllHttpClientNames(ctx.ServiceProvider))
    {
        var pipelineKey = Helper.GetPipelineKey(ctx.ServiceProvider, clientName);
        ctx.AddResiliencePipeline<TKey, TValue>(pipelineKey, (builder, context) => { /* ... */ });
    }
});
```

2) Second problem is adding a ResilienceHandler to the message handlers chain, as it needs a pipeline name/key to be constructed:

```c#
_ = builder.AddHttpMessageHandler(serviceProvider =>
{
    var selector = CreatePipelineSelector(serviceProvider, pipelineBuilder.PipelineName);
    var provider = serviceProvider.GetRequiredService<ResiliencePipelineProvider<HttpKey>>();

    return new ResilienceHandler(selector);
});
```

By the time the callback provided to AddHttpMessageHandler is executed, client's name is known already, but it's not available to the callback. _This was what I initially thought this issue is about._

_As much as I'd hate suggesting this_, but, as a workaround while a better API is not available, you can try using configuration methods/approaches that give you access to the full `HttpMessageHandlerBuilder` instance (which has a `Name` property assigned), such as `ConfigureHttpMessageHandlerBuilder` (we ~~tried to~~ deprecated this one), `IHttpMessageHandlerBuilderFilter` or even directly `HttpClientFactoryOptions.HttpMessageHandlerBuilderActions`

```c#
builder.ConfigureHttpMessageHandlerBuilder(b =>
{
    var pipelineName = Helper.GetPipelineName(b.Services, b.Name);
    // ...
    b.AdditionalHandlers.Add(/* ... */);
});
```

----------

I was thinking to change this issue to be about having the client Name available inside callbacks like AddHttpMessageHandler, but decided it's better to file it separately, as it would be an API proposal.

I will now close this issue as "by design", but please let me know if the provided workarounds won't work for you for some reason. Thanks!

**Comment by iliar-turdushev** [comment]
@CarnaViire Thank you.
